pkg/standalone-utils/contracts/BalancerSorQueries.sol

```diff
@@ -0,0 +1,293 @@
+        returns (
+            uint256[][] memory balances,
+            uint256[] memory totalSupplies,
+            uint256[] memory swapFees,
+            uint256[] memory linearWrappedTokenRates,
+            uint256[][] memory weights,
+            uint256[][] memory scalingFactors,
+            uint256[] memory amps
+        )
+    {
+        uint256 i;
+        address[] memory pools = new address[](poolIds.length);
+
+        for (i = 0; i < poolIds.length; i++) {
+            (pools[i], ) = vault.getPool(poolIds[i]);
+        }
+
+        if (config.loadTokenBalanceUpdatesAfterBlock) {
+            balances = getPoolTokenBalancesWithUpdatesAfterBlock(poolIds, config.blockNumber);
+        }
+
+        if (config.loadTotalSupply) {
+            totalSupplies = getTotalSupplyForPools(pools, config.totalSupplyTypes);
+        }
+
+        if (config.loadSwapFees) {
+            swapFees = getSwapFeePercentageForPools(pools, config.swapFeeTypes);
+        }
+
+        if (config.loadLinearWrappedTokenRates) {
+            address[] memory linearPools = new address[](config.linearPoolIdxs.length);
+
+            for (i = 0; i < config.linearPoolIdxs.length; i++) {
+                linearPools[i] = pools[config.linearPoolIdxs[i]];
+            }
+
+            linearWrappedTokenRates = getWrappedTokenRateForLinearPools(linearPools);
+        }
+
+        if (config.loadNormalizedWeights) {
+            address[] memory weightedPools = new address[](config.weightedPoolIdxs.length);
+
+            for (i = 0; i < config.weightedPoolIdxs.length; i++) {
+                weightedPools[i] = pools[config.weightedPoolIdxs[i]];
+            }
+
+            weights = getNormalizedWeightsForPools(weightedPools);
+        }
+
+        if (config.loadScalingFactors) {
+            address[] memory scalingFactorPools = new address[](config.scalingFactorPoolIdxs.length);
+
+            for (i = 0; i < config.scalingFactorPoolIdxs.length; i++) {
+                scalingFactorPools[i] = pools[config.scalingFactorPoolIdxs[i]];
+            }
+
+            scalingFactors = getScalingFactorsForPools(scalingFactorPools);
+        }
+
+        if (config.loadAmps) {
+            address[] memory ampPools = new address[](config.ampPoolIdxs.length);
+
+            for (i = 0; i < config.ampPoolIdxs.length; i++) {
+                ampPools[i] = pools[config.ampPoolIdxs[i]];
+            }
+
+            amps = getAmpForPools(ampPools);
+        }
+    }
+
+    function getPoolTokenBalancesWithUpdatesAfterBlock(bytes32[] memory poolIds, uint256 blockNumber)
+        public
+        view
+        returns (uint256[][] memory)
+    {
+        uint256[] memory balances;
+        uint256 lastChangeBlock;
+        uint256[][] memory allBalances = new uint256[][](poolIds.length);
+
+        for (uint256 i = 0; i < poolIds.length; i++) {
+            (, balances, lastChangeBlock) = vault.getPoolTokens(poolIds[i]);
+
+            if (lastChangeBlock > blockNumber) {
```

Comment:
Is the idea here to just limit the rpc calldata size? Since the `getPoolTokens` call is made regardless it seems better to always update that data. The onchain node state is the source of truth. An example of this is right now on Polygon the subgraph is up to the latest block but has slightly bad/inaccurate data. This means the onchain data doesn't overwrite any existing bad subgraph data

Reply:
Hey sorry for the delay here, missed the notifications.

The overall goal of this contract was to limit both data in and data out, and the token balances represent the largest amount of data out for the query. The initial assumption was that the subgraph balances should be correct (until the last block processed), so only loading changes after that block would significantly reduce the data out. This could also extend to polling queries (ie: if you're fetching every 30-60 seconds, you really dont need all balances on every fetch).

You can always pass a `blockNumber` of 0 and bypass the check entirely, I added this to the config in the SOR, it previously was not there. Definitely not stuck on keeping it though, but it does produce meaningful reduction in the query time simply by reducing the amount of data being transferred across the network.